reconcile/openshift_tekton_resources.py

```diff
@@ -0,0 +1,353 @@
+import sys
+import logging
+import json
+from typing import Any, Iterable, Optional, Union
+
+import yaml
+import jinja2
+
+from reconcile import queries
+from reconcile import openshift_base as ob
+from reconcile.status import ExitCodes
+from reconcile.utils import gql
+from reconcile.utils.defer import defer
+from reconcile.utils.openshift_resource import OpenshiftResource as OR
+from reconcile.utils.semver_helper import make_semver
+
+LOG = logging.getLogger(__name__)
+QONTRACT_INTEGRATION = 'openshift-tekton-resources'
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
+
+# it must be a single character due to resource max length
+OBJECTS_PREFIX = 'o'
+RESOURCE_MAX_LENGTH = 63
+
+# Defaults
+DEFAULT_DEPLOY_RESOURCES = {'requests': {'cpu': '50m',
+                                         'memory': '200Mi'},
+                            'limits': {'cpu': '200m',
+                                       'memory': '300Mi'}}
+SAAS_FILES_QUERY = """
+{
+  saas_files: saas_files_v2 {
+    path
+    name
+    configurableResources
+    pipelinesProvider {
+      name
+      provider
+    }
+    deployResources {
+      requests {
+        cpu
+        memory
+      }
+      limits {
+        cpu
+        memory
+      }
+    }
+  }
+}
+"""
+
+
+class OpenshiftTektonResourcesNameTooLongError(Exception):
+    pass
+
+
+class OpenshiftTektonResourcesBadConfigError(Exception):
+    pass
+
+
+# Builds a list of v2 saas files from qontract-server
+def get_saas_files(saas_file_name: Optional[str]) -> list[dict[str, Any]]:
+    saas_files = [
+        s for s in gql.get_api().query(SAAS_FILES_QUERY)['saas_files']
+        if s.get('configurableResources')]
+
+    if saas_file_name:
+        saas_file = None
+        for s in saas_files:
+            if s['name'] == saas_file_name:
+                saas_file = s
+                break
+
+        return [saas_file] if saas_file else []
+
+    return saas_files
+
+
+def fetch_tkn_providers(saas_files: Iterable[dict[str, Any]]) \
+        -> dict[str, Any]:
+    duplicates: set[str] = set()
+    all_tkn_providers = {}
+    for pipeline_provider in queries.get_pipelines_providers():
+        if pipeline_provider['provider'] != 'tekton':
+            continue
+
+        if pipeline_provider['name'] in all_tkn_providers:
+            duplicates.add(pipeline_provider['name'])
+        else:
+            all_tkn_providers[pipeline_provider['name']] = pipeline_provider
+
+    if duplicates:
+        raise OpenshiftTektonResourcesBadConfigError(
+            'There are duplicates in tekton providers names: '
+            f'{", ".join(duplicates)}')
+
+    # Only get the providers that are used by the saas files
+    # Add the saas files belonging to it
+    tkn_providers = {}
+    for saas_file in saas_files:
```

Comment:
Please, please, please, use loop variables that are very different from the iterated object. `sf` works for me. I've made too many mistakes with these too-similar identifiers.

Reply:
fixed in aad5bfe